server/tso.go

```diff
@@ -116,23 +112,26 @@ func (s *Server) syncTimestamp() error {
 }
 
 func (s *Server) updateTimestamp() error {
-	prev := s.ts.Load().(*atomicObject)
+	prev := s.ts.Load().(*atomicObject).physical
 	now := time.Now()
 
-	since := now.Sub(prev.physical).Nanoseconds() / 1e6
+	since := now.Sub(prev)
 	if since > 3*updateTimestampStep {
-		log.Warnf("clock offset: %v, prev: %v, now %v", since, prev.physical, now)
+		log.Warnf("clock offset: %v, prev: %v, now: %v", since, prev, now)
 	}
 	// Avoid the same physical time stamp
 	if since <= 0 {
-		log.Warnf("invalid physical timestamp, prev:%v, now:%v, re-update later", prev.physical, now)
+		log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
 		return nil
 	}
 
-	if now.Sub(s.lastSavedTime).Nanoseconds()/1e6 > s.cfg.TsoSaveInterval {
-		if err := s.saveTimestamp(now); err != nil {
+	if now.Sub(s.lastSavedTime) >= 0 {
+		save := now.Add(s.cfg.TsoSaveInterval.Duration)
+		if err := s.saveTimestamp(save); err != nil {
 			return errors.Trace(err)
 		}
+
+		log.Debugf("save timestamp ok: prev %v save %v", prev, save)
```

Comment:
So how about define a variable for `s.ts.Load().(*atomicObject)`, then we can trace this time in log.